pkg/backup/backuper.go

```diff
@@ -0,0 +1,321 @@
+package backup
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	"github.com/mattn/go-sqlite3"
+
+	"github.com/pkg/errors"
+)
+
+// BackupFilenamePrefix is the prefix used in every backup file.
+const BackupFilenamePrefix = "tbl_backup"
+
+var (
+	open          = openDatabase
+	newBackupFile = createBackupFile
+)
+
+// Backuper is the process that executes the backup process.
+type Backuper struct {
+	source DB
+	result *BackupResult
+	config *Config
+}
+
+// NewBackuper creates a new backuper responsible for making backups of a SQLite database.
+func NewBackuper(sourcePath string, backupDir string, opts ...Option) (*Backuper, error) {
+	config := DefaultConfig()
+	for _, o := range opts {
+		if err := o(config); err != nil {
+			return nil, err
+		}
+	}
+
+	b := &Backuper{
+		result: &BackupResult{
+			Dir: backupDir,
+		},
+		config: config,
+	}
+
+	if err := b.init(sourcePath, backupDir); err != nil {
+		return nil, err
+	}
+
+	return b, nil
+}
+
+// Backup creates a backup to a temporary in disk file.
+func (b *Backuper) Backup(ctx context.Context) (*BackupResult, error) {
+	// SQLite backup starts
+	startTime := time.Now()
+
+	connA, err := b.source.Conn(ctx)
+	if err != nil {
+		return b.result, errors.Errorf("getting db conn: %s", err)
+	}
+	defer func() {
+		_ = connA.Close()
+	}()
+
+	connB, err := b.result.Database.Conn(ctx)
+	if err != nil {
+		return b.result, errors.Errorf("getting backup db conn: %s", err)
+	}
+	defer func() {
+		_ = connB.Close()
+	}()
+
+	if err := b.doBackup(connA, connB); err != nil {
+		return b.result, errors.Errorf("backup: %s", err)
```

Comment:
I don't return `BackupResult{}` because `b.result` contains the partially backuped file. Your comment below is a good suggestion and solves this issues by decouple the result from the information where the backup file is.